Lombiq.EmailClient/Services/ImapEmailClient.cs

```diff
@@ -0,0 +1,237 @@
+using Lombiq.EmailClient.Constants;
+using Lombiq.EmailClient.Models;
+using MailKit;
+using MailKit.Net.Imap;
+using MailKit.Search;
+using Microsoft.Extensions.Options;
+using MimeKit;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using UniqueId = MailKit.UniqueId;
+
+namespace Lombiq.EmailClient.Services;
+
+public class ImapEmailClient : IEmailClient
+{
+    private readonly Dictionary<uint, MimeMessage> _downloadedMessages = [];
+    private readonly ImapSettings _imapSettings;
+    private ImapClient _imapClient;
+    private bool _isDisposed;
+
+    public ImapEmailClient(IOptionsSnapshot<ImapSettings> imapSettings) =>
+        _imapSettings = imapSettings.Value;
+
+    public async Task<IEnumerable<EmailMessage>> GetEmailsAsync(EmailFilterParameters parameters)
+    {
+        ArgumentNullException.ThrowIfNull(parameters);
+
+        await InitializeImapClientAndConnectIfNeededAsync();
+        var folder = await OpenFolderAsync();
+
+        var searchQuery = BuildSearchQuery(parameters);
+
+        var uids = (await folder.SearchAsync(searchQuery ?? SearchQuery.All)).Select(uniqueId => uniqueId.Id).ToList();
+
+        // Not using UID range search if we're on localhost as it's not supported by some dev servers. UIDs are filtered
+        // in memory.
+        uids = IsLocalhost() ? uids.Where(uid => uid > parameters.AfterImapUniqueId).ToList() : uids;
+        if (uids.Count == 0) return [];
+
+        var messages = new List<EmailMessage>();
+        foreach (var uid in uids)
+        {
+            var mimeMessage = await GetOrDownloadMessageByUniqueIdAsync(uid, folder);
+            if (mimeMessage == null) continue;
+
+            messages.Add(MapToEmailMessage(mimeMessage, folder.FullName, uid));
+        }
+
+        return messages;
+    }
+
+    public async Task<Stream> GetAttachmentStreamAsync(
+        EmailMessage emailMessage,
+        AttachmentMetadata attachmentMetadata)
+    {
+        ArgumentNullException.ThrowIfNull(emailMessage);
+        ArgumentNullException.ThrowIfNull(attachmentMetadata);
+
+        ValidateImapProtocol(emailMessage);
+        var message = await GetOrDownloadMessageByUniqueIdAsync(emailMessage.Metadata.ImapUniqueId);
+
+        var attachmentEntity = message.Attachments.FirstOrDefault(attachment =>
+            attachment.ContentDisposition.FileName == attachmentMetadata.FileName);
```

Comment:
Since these are file names I have to ask: is this intentionally a case sensitive check?

Reply:
Didn't think about it, but theoretically, it's possible to attach littlekitty.png and LittleKitty.png to the same email if the email client allows it therefore we can be in a situation where IMAP retrieves both.